lib/puppet/provider/exec/pwsh.rb

```diff
@@ -51,6 +42,71 @@ def validatecmd(command)
   end
 
   private
+
+  def get_pwsh_command
+    if @resource['path'].nil?
+      pwsh = Puppet::Util.which('pwsh')
+    else
+      pwsh = which_with_custom_env('pwsh', @resource['path'])
+    end
+    return pwsh unless pwsh.nil?
+
+    return nil unless Puppet::Util::Platform.windows?
+    # These paths are Windows only
+    pwsh = "#{ENV['ProgramFiles']}\\PowerShell\\6\\pwsh.exe"
+    return pwsh if File.exists?(pwsh)
+    pwsh = "#{ENV['ProgramFiles(x86)']}\\PowerShell\\6\\pwsh.exe"
+    File.exists?(pwsh) ? pwsh : nil
+  end
+
+  # Based on which command from https://github.com/puppetlabs/puppet/blob/1c14d0a9fdfc31933603e571b616b6cd675e6b71/lib/puppet/util.rb#L241-L286
+  # Resolve a path for an executable to the absolute path. This tries to behave
+  # in the same manner as the unix `which` command and uses the `PATH`
+  # environment variable.
```

Comment:
I don't like that we're reimplementing code that already exists in Puppet. Couldn't you just wrap the existing Puppet call using withenv? https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util.rb#L119-L146

Usage in the existing `exec` is something like this - 

```ruby
if resource[:path]
  Puppet::Util.withenv( {'PATH' => resource[:path].join(File::PATH_SEPARATOR)}, :windows) do
    return if which(exe)
  end
end
```

https://github.com/puppetlabs/puppet/blob/ba4d1a1aba0095d3c70b98fea5c67434a4876a61/lib/puppet/provider/exec/windows.rb#L47-L51

Reply:
I don't like it either but that method is marked `private` so I shouldn't be calling it.  I could monkey patch a public method but that seems even worse.

Given that it's a "fairly" static method I decided to copy and modify instead.

Reply:
Both `withenv` at https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util.rb#L119-L135 and `which` at https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util.rb#L241-L286 are module functions exposed on `Puppet::Util`.

Not sure I understand?

Reply:
You are right...apologies

Reply:
With your code example above.  That will work, it just seems heavy handed to change the search path temporarily instead of just modifying the which command.

Also while _probably_ not a problem for MRI ruby, it wouldn't be thread safe for real threads.

Reply:
This is the pattern we've established in Puppet itself, so it's preferable to stay consistent... thread safety isn't a concern here.